packages/flutter_tools/lib/src/android/java.dart

```diff
@@ -0,0 +1,245 @@
+
+  /// The path of the runtime's java binary.
+  ///
+  /// This should be only used for logging and validation purposes.
+  /// If you need to invoke the binary directly, consider adding a new method
+  /// to this class instead.
+  final String binary;
+
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final OperatingSystemUtils _os;
+  final Platform _platform;
+  final ProcessManager _processManager;
+  final ProcessUtils _processUtils;
+
+  JavaVersion? _version;
+
+  /// Returns an environment variable map with
+  /// 1. JAVA_HOME set if this object has a known home directory, and
+  /// 2. The java binary folder appended onto PATH, if the binary location is known.
+  ///
+  /// This map should be used as the environment when invoking any Java-dependent
+  /// processes, such as Gradle or Android SDK tools (avdmanager, sdkmanager, etc.)
+  Map<String, String> getJavaEnvironment() {
+    return <String, String>{
+      if (home != null) _javaHomeEnvironmentVariable: home!,
+      'PATH': _fileSystem.path.dirname(binary) +
+                        _os.pathVarSeparator +
+                        _platform.environment['PATH']!,
+    };
+  }
+
+  /// Returns the version of java in the format \d(.\d)+(.\d)+
+  /// Returns null if version could not be determined.
+  JavaVersion? getVersion() {
+    if (_version == null) {
+      final RunResult result = _processUtils.runSync(
+        <String>[binary, '--version'],
+        environment: getJavaEnvironment(),
+      );
+      if (result.exitCode != 0) {
+        _logger.printTrace('java --version failed: exitCode: ${result.exitCode}'
+          ' stdout: ${result.stdout} stderr: ${result.stderr}');
+      }
+      _version = JavaVersion.tryParseFromJavaOutput(result.stdout, logger: _logger);
+    }
+
+    return _version;
+  }
+
+  bool canRun() {
+    return _processManager.canRun(binary);
+  }
+}
+
+String? _findJavaHome({
+  required Logger logger,
+  required AndroidStudio? androidStudio,
+  required Platform platform,
+}) {
+  final String? androidStudioJavaPath = androidStudio?.javaPath;
+  if (androidStudioJavaPath != null) {
+    logger.printTrace("Using Android Studio's java.");
+    return androidStudioJavaPath;
+  }
+
+  final String? javaHomeEnv = platform.environment[_javaHomeEnvironmentVariable];
+  if (javaHomeEnv != null) {
+    logger.printTrace('Using JAVA_HOME from environment valuables.');
+    return javaHomeEnv;
+  }
+  return null;
+}
+
+String? _findJavaBinary({
+  required Logger logger,
+  required String? javaHome,
+  required FileSystem fileSystem,
+  required OperatingSystemUtils operatingSystemUtils,
+  required Platform platform,
+}) {
+  try {
+    if (javaHome != null) {
+      return fileSystem.path.join(javaHome, 'bin', 'java');
+    }
+
+    // Fallback to PATH based lookup.
+    final String? pathJava = operatingSystemUtils.which(_kJavaExecutable)?.path;
+    if (pathJava != null) {
+      logger.printTrace('Using java from PATH.');
```

Comment:
These are potentially useful when triaging issues like https://github.com/flutter/flutter/issues/124776 where the user has multiple java installs and we want to see which one flutter is using.

However, `flutter doctor -v` does this for us, and will be more reliable once the parent issue is solved, so I would not be opposed to deleting these traces or at least moving them to callers.

Reply:
There's a larger discussion about removing a lot of our traces, since our current verbose test logs are almost impossible to parse. If `flutter doctor -v` gives us the right answer, let's rely on that, no need to duplicate.

Reply:
On the larger discussion of removing traces, I'm not against adding traces altogether, just want to point out that they're not free, and the amount of traces we have in devicelab logs can be considered tech debt.

Reply:
Removed non-warning prints.